source/server/server.cc

```diff
@@ -364,23 +358,37 @@ void InstanceImpl::run() {
   guard_dog_->stopWatching(watchdog);
   watchdog.reset();
 
+  terminate();
+}
+
+void InstanceImpl::terminate() {
```

Comment:
You might consider having someone at Lyft smoke test this change when ready. The shutdown stuff is tricky to get right and it's very difficult to visually review for errors.